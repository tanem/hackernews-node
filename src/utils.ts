import jwt from 'jsonwebtoken'
import { Context } from './context'

export const APP_SECRET = 'GraphQL-is-aw3some'

export const getUserId = (context: Context) => {
  const Authorization = context.request.get('Authorization')
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '')
    const { userId } = jwt.verify(token, APP_SECRET) as { userId: number }
    return userId
  }

  throw new Error('Not authenticated')
}
