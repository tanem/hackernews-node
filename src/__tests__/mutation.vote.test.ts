import { Link, User } from '@prisma/client'
import bcrypt from 'bcryptjs'
import Chance from 'chance'
import jwt from 'jsonwebtoken'
import path from 'path'
import request from 'supertest'
import { APP_SECRET } from '../utils'

const chance = Chance(path.basename(__filename))

let httpServerUrl: typeof global.__HTTP_SERVER_URL__
let prisma: typeof global.__PRISMA__

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()
const description = chance.sentence({ words: 5 })
const url = chance.url()

let linkId: Link['id']
let token: string
let userId: User['id']

beforeEach(async () => {
  ;({ id: userId } = await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
    },
  }))
  ;({ id: linkId } = await prisma.link.create({
    data: {
      description,
      url,
      postedBy: {
        connect: { id: userId },
      },
    },
  }))
  token = jwt.sign({ userId }, APP_SECRET)
})

test('new vote', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .set('Authorization', `Bearer ${token}`)
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
            link {
              description
              id
              url
            }
            user {
              email
              id
              name
            }
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('duplicate vote', async () => {
  await prisma.vote.create({
    data: {
      link: {
        connect: { id: linkId },
      },
      user: {
        connect: { id: userId },
      },
    },
  })

  const { body, status } = await request(httpServerUrl)
    .post('/')
    .set('Authorization', `Bearer ${token}`)
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('not authenticated', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
        mutation {
          vote(linkId: "${linkId}") {
            id
            link {
              description
              id
              url
            }
            user {
              email
              id
              name
            }
          }
        }
      `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})
