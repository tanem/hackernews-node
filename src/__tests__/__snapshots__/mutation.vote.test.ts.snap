// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicate vote 1`] = `
Object {
  "data": Object {
    "vote": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Already voted for link: 1",
      "path": Array [
        "vote",
      ],
    },
  ],
}
`;

exports[`new vote 1`] = `
Object {
  "data": Object {
    "vote": Object {
      "id": 1,
      "link": Object {
        "description": "Nem nokibof bar jilhe puwa.",
        "id": 1,
        "url": "http://gobhaptem.mq/ki",
      },
      "user": Object {
        "email": "juhiza@sapeag.zw",
        "id": 1,
        "name": "Dollie",
      },
    },
  },
}
`;

exports[`not authenticated 1`] = `
Object {
  "data": Object {
    "vote": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Not authenticated",
      "path": Array [
        "vote",
      ],
    },
  ],
}
`;
